va/tlsalpn.go

```diff
@@ -150,34 +183,55 @@ func (va *ValidationAuthorityImpl) getChallengeCert(
 	cs := conn.(*tls.Conn).ConnectionState()
 	certs := cs.PeerCertificates
 	if len(certs) == 0 {
-		va.log.Infof("%s challenge for %s resulted in no certificates", core.ChallengeTypeTLSALPN01, identifier.Value)
+		va.log.Infof("%s challenge for %s resulted in no certificates", core.ChallengeTypeTLSALPN01, ident.Value)
 		return nil, nil, berrors.UnauthorizedError("No certs presented for %s challenge", core.ChallengeTypeTLSALPN01)
 	}
 	for i, cert := range certs {
 		va.log.AuditInfof("%s challenge for %s received certificate (%d of %d): cert=[%s]",
-			core.ChallengeTypeTLSALPN01, identifier.Value, i+1, len(certs), hex.EncodeToString(cert.Raw))
+			core.ChallengeTypeTLSALPN01, ident.Value, i+1, len(certs), hex.EncodeToString(cert.Raw))
 	}
 	return certs[0], &cs, nil
 }
 
-func checkExpectedSAN(cert *x509.Certificate, name identifier.ACMEIdentifier) error {
-	if len(cert.DNSNames) != 1 {
-		return errors.New("wrong number of dNSNames")
+func checkExpectedSAN(cert *x509.Certificate, ident identifier.ACMEIdentifier) error {
+	var asn1Tag int
+	var certSAN []byte
+	switch ident.Type {
+	case identifier.TypeDNS:
+		if len(cert.DNSNames) != 1 || len(cert.IPAddresses) != 0 {
+			return errors.New("wrong number of identifiers")
+		}
+		asn1Tag = 2
+		certSAN = []byte(cert.DNSNames[0])
+	case identifier.TypeIP:
+		if len(cert.IPAddresses) != 1 || len(cert.DNSNames) != 0 {
+			return errors.New("wrong number of identifiers")
+		}
+		asn1Tag = 7
+		certSAN = cert.IPAddresses[0]
+	default:
+		// This should never happen. The calling function should check the
+		// identifier type.
+		return fmt.Errorf("Unknown identifier type: %s", ident.Type)
 	}
 
 	for _, ext := range cert.Extensions {
 		if IdCeSubjectAltName.Equal(ext.Id) {
 			expectedSANs, err := asn1.Marshal([]asn1.RawValue{
-				{Tag: 2, Class: 2, Bytes: []byte(cert.DNSNames[0])},
+				{Tag: asn1Tag, Class: 2, Bytes: certSAN},
 			})
```

Comment:
I know you're not touching this, but I think this function would be much clearer if it had the structure:
```
switch ident.Type {
case DNS:
  expectedSANBytes := asn1.Marshal(RawValue{Tag: 2...
  expectedSAN := cert.DNSNames[0]
case IP:
  expectedSANBytes := asn1.Marshal(RawValue{Tag: 7...
  expectedSAN := []byte(net.IP[cert.IPAddresses[0].String()...
}

compare expectedSANBytes and real extension bytes

compare expectedSAN and ident.Value
```

This has three improvements:
1) no need for a separate asn1Tag variable
2) no need for the asn1-marshalling logic to be inside the loop
3) no need for an IP special-case before the EqualFold